src/Model/Aggregate.php

```diff
@@ -0,0 +1,294 @@
+
+            // now add the expressions here
+            $this->addExpression($name, $seed);
+        }
+
+        return $this;
+    }
+
+    public function getRef(string $link): Reference
+    {
+        return $this->baseModel->getRef($link);
+    }
+
+    /**
+     * Method to enable commutative usage of methods enabling both of below
+     * Resulting in Aggregate on $model.
+     *
+     * $model->groupBy(['abc'])->withAggregateField('xyz');
+     *
+     * and
+     *
+     * $model->withAggregateField('xyz')->groupBy(['abc']);
+     */
+    public function withAggregateField(string $name, $seed = []): Model
+    {
+        static::addField($name, $seed);
+
+        return $this;
+    }
+
+    /**
+     * Adds new field into model.
+     *
+     * @param array|object $seed
+     */
+    public function addField(string $name, $seed = []): Field
+    {
+        $seed = is_array($seed) ? $seed : [$seed];
+
+        if (isset($seed[0]) && $seed[0] instanceof FieldSqlExpression) {
+            return parent::addField($name, $seed[0]);
+        }
+
+        if ($seed['never_persist'] ?? false) {
+            return parent::addField($name, $seed);
+        }
+
+        if ($this->baseModel->hasField($name)) {
+            $field = clone $this->baseModel->getField($name);
+            $field->unsetOwner(); // will be new owner
+        } else {
+            $field = null;
+        }
+
+        return $field
+            ? parent::addField($name, $field)->setDefaults($seed)
+            : parent::addField($name, $seed);
+    }
+
+    public function setLimit(int $count = null, int $offset = 0)
+    {
+        $this->baseModel->setLimit($count, $offset);
+
+        return $this;
+    }
+
+    /**
+     * @param string $mode
+     * @param array  $args
+     *
+     * @return Query
+     */
+    public function action($mode, $args = [])
+    {
+        switch ($mode) {
+            case 'select':
+                $fields = $this->only_fields ?: array_keys($this->getFields());
+
+                // select but no need your fields
+                $query = $this->baseModel->action($mode, [false]);
+
+                $this->initQueryFields($query, array_unique($fields + $this->groupByFields));
+                $this->initQueryOrder($query);
+                $this->initQueryGrouping($query);
+                $this->initQueryConditions($query);
+
+                $this->hook(self::HOOK_INIT_SELECT_QUERY, [$query]);
+
+                return $query;
+            case 'count':
+                $query = $this->baseModel->action($mode, $args);
+
+                $query->reset('field')->field($this->expr('1'));
+                $this->initQueryGrouping($query);
+
+                $this->hook(self::HOOK_INIT_SELECT_QUERY, [$query]);
+
+                return $query->dsql()->field('count(*)')->table($this->expr('([]) der', [$query]));
+            case 'field':
+            case 'fx':
```

Comment:
Not sure if `field` should be allowed and also how `fx` works. Need to see some usecase example in SQL and tests.